Работа по спринту 19

UtinsStory opened the pull request.

UtinsStory committed:
Сверстаны экраны запуска и главный экран

UtinsStory committed:
Переработка кода с учетом необходимости использования NavigationStack

UtinsStory committed:
Реализация функциона согласно заданию спринта. Необходимо добавить разделительную линию над таббаром

UtinsStory committed:
Разделитель у таббар отображается корректно. Добавлены моковые логотипы перевозчиков

NaR0RoW commented on TravelSchedule/TravelSchedule.xcodeproj/project.xcworkspace/xcuserdata/genautin.xcuserdatad/UserInterfaceState.xcuserstate:
Папку `xcuserdata` стоит игнорировать, кладя в `.gitignore`. В ней хранится информация о состоянии проекта (какой файл открыт, какой таргет выбран и т.п.), брейкпоинты и схемы для текущего пользователя Xcode.
Про `gitignore` можно прочитать вот [тут](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files)

NaR0RoW commented on TravelSchedule/TravelSchedule/Assets.xcassets/Images/ReverseButton.imageset/Сhange.png:
Картинки 1x лучше не добавлять. Они нужны были для iPhone без retina экранов, а последний такой - iPhone5. Он не поддерживает минимальную версию учебного приложения, поэтому можно смело удалять, чтобы не занимать лишнюю память

NaR0RoW commented on TravelSchedule/TravelSchedule/Helpers/CustomBackButton.swift:
```diff
@@ -0,0 +1,21 @@
+
+    var body: some View {
+        Button(action: {
+            self.presentationMode.wrappedValue.dismiss()
```
`self` везде лучше опускать, если явно того не требуется

NaR0RoW commented on TravelSchedule/TravelSchedule/Helpers/CustomBackButton.swift:
```diff
@@ -0,0 +1,21 @@
+import SwiftUI
+
+struct CustomBackButton: View {
+    @Environment(\.presentationMode) var presentationMode: Binding<PresentationMode>
```
Лучше сделать поле приватным

NaR0RoW commented on TravelSchedule/TravelSchedule/Helpers/Mock.swift:
```diff
@@ -0,0 +1,60 @@
+        ]
+    }()
+    
+    static let cities: [City] = [
+            City(title: "Москва", stations: [
+                Station(title: "Курский Вокзал"),
+                Station(title: "Казанский Вокзал"),
+                Station(title: "Киевский Вокзал"),
+                Station(title: "Ленинградский Вокзал")
+            ]),
+            City(title: "Санкт-Петербург", stations: [
+                Station(title: "Московский Вокзал"),
+                Station(title: "Витебский Вокзал"),
+                Station(title: "Ладожский Вокзал"),
+                Station(title: "Балтийский Вокзал")
+            ]),
+            City(title: "Сочи", stations: [
+                Station(title: "Вокзал Сочи"),
+                Station(title: "Вокзал Адлер"),
+                Station(title: "Вокзал Лазаревская")
+            ]),
+            City(title: "Краснодар", stations: [
+                Station(title: "Краснодар I"),
+                Station(title: "Краснодар II")
+            ])
+            ]
```
Форматирование кода немного поехало, можно выделить это место и воспользоваться автоматическим форматированием в Xcode `control + I`

NaR0RoW commented on TravelSchedule/TravelSchedule/Model/CarrierModel.swift:
```diff
@@ -0,0 +1,15 @@
+
+struct CarrierModel: Hashable {
+    let name: String
+    let logo: String
```
Когда видишь слово `logo`, то ожидаешь увидеть константу типа `Image`, но не `String`. Лучше поменять название константы на `logoName`

NaR0RoW commented on TravelSchedule/TravelSchedule/Model/CityModel.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+struct City: Identifiable {
```
Название структуры не соот. названию файла и хедера файла. Лучше придерживаться констистентности

NaR0RoW commented on TravelSchedule/TravelSchedule/Model/StationModel.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+struct Station: Identifiable, Hashable {
```
Тут тоже не совпадает с названием файла

NaR0RoW commented on TravelSchedule/TravelSchedule/Model/TripModel.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+struct Trip: Identifiable {
```
Тут тоже не совпадает с названием файла. Перепроверь пожалуйста весь код. Будет лучше, если название сущности совпадает с названием файла, в котором эта сущность лежит

NaR0RoW commented on TravelSchedule/TravelSchedule/TravelScheduleApp.swift:
```diff
@@ -9,9 +9,12 @@ import SwiftUI
 
 @main
 struct TravelScheduleApp: App {
+    @AppStorage("isDarkMode") var isDarkMode: Bool = false
```
Перепроверь все свойства и методы в классах. Они должны быть максимально инкапсулированы и скрыты внутри класса. Используй модификатор `private` для достижения этой цели.

NaR0RoW replied:
Название ключа можно в отдельную сущность вынести

NaR0RoW commented on TravelSchedule/TravelSchedule/View/CarrierInfoView.swift:
```diff
@@ -0,0 +1,68 @@
+import SwiftUI
+
+struct CarrierInfoView: View {
+    var carrier: CarrierModel
```
Лучше константой сделать

NaR0RoW commented on TravelSchedule/TravelSchedule/View/CarrierInfoView.swift:
```diff
@@ -0,0 +1,68 @@
+    
+    var body: some View {
+        ZStack {
+            Color.ypWhite.edgesIgnoringSafeArea(.all)
```
Лучше ` .ignoresSafeArea()` использовать `.edgesIgnoringSafeArea` больше не актуально

NaR0RoW commented on TravelSchedule/TravelSchedule/View/CityListView.swift:
```diff
@@ -0,0 +1,60 @@
+
+struct CityListView: View {
+    @State private var searchString = ""
+    @StateObject var viewModel = StationsViewModel()
```
`@StateObject` всегда должен быть приватным

NaR0RoW commented on TravelSchedule/TravelSchedule/View/CityListView.swift:
```diff
@@ -0,0 +1,60 @@
+struct CityListView: View {
+    @State private var searchString = ""
+    @StateObject var viewModel = StationsViewModel()
+    var selectAction: (String) -> Void
```
Тоже лучше константу использовать. Перепроверь пожалуйста весь код. Везде где можно, лучше константы использовать, а не переменные

NaR0RoW commented on TravelSchedule/TravelSchedule/View/CityListView.swift:
```diff
@@ -0,0 +1,60 @@
+    var body: some View {
+        VStack {
+            SearchBarView(searchText: $searchString)
+            if filteredCities.isEmpty {
```
Везде, где это возможно, лучше использовать `opacity` вместо `if` для сокрытия `View`. [Вот хороший пример с ссылками на `WWDC`](https://stackoverflow.com/a/57420479)

NaR0RoW commented on TravelSchedule/TravelSchedule/View/MainScreenView.swift:
```diff
@@ -0,0 +1,57 @@
+                UITabBar.appearance().backgroundColor = .ypWhite
+            }
+            .tint(.ypBlack)
+            .onChange(of: showTabBar) { newValue in
```
Лучше везде осмысленные названия использовать вместо `newValue`. Например, в данном случае возвращаемое значение можно назвать: `isShown`

NaR0RoW commented on TravelSchedule/TravelSchedule/View/ScheduleView.swift:
```diff
@@ -0,0 +1,158 @@
+    @State private var toStation = ""
+    @State private var path: [Destination] = []
+    @StateObject var citiesViewModel = StationsViewModel()
+    @StateObject var tripsViewModel = TripsViewModel(carriersViewModel: CarriersViewModel())
```
Одна  `ObservableObject` `ViewModel` используется внутри другой `ObservableObject` `ViewModel`. Кажется `@Published` поля вложенной модели не будут обновляться.
Это не очень хороший паттерн. Вот [статья](https://rhonabwy.com/2021/02/13/nested-observable-objects-in-swiftui/) на этот счет

NaR0RoW commented on TravelSchedule/TravelSchedule/View/SearchBarView.swift:
```diff
@@ -0,0 +1,72 @@
+    @State var isSearching: Bool = false
+    var placeholder = "Введите запрос"
+    
+    var body: some View {
+            HStack (spacing: 0) {
```
В этой `View` форматирование кода поехало

NaR0RoW commented on TravelSchedule/TravelSchedule/View/SearchBarView.swift:
```diff
@@ -0,0 +1,72 @@
+
+struct SearchBarView: View {
+    @Binding var searchText: String
+    @State var isSearching: Bool = false
```
Все `@State` свойства рекомендую указывать как `private`, потому что они не должны быть видны за пределами `View`. Если нужно изменить значение внутри `View`, то тогда следует использовать `@Binding`

NaR0RoW commented on TravelSchedule/TravelSchedule/View/SettingsView.swift:
```diff
@@ -0,0 +1,49 @@
+
+struct SettingsView: View {
+    @Binding var isDarkMode: Bool
+    let termsOfServiceURLString = "https://yandex.ru/legal/practicum_offer/"
```
Можно вынести в отдельную сущность с константами

NaR0RoW commented on TravelSchedule/TravelSchedule/View/TripFilterView.swift:
```diff
@@ -0,0 +1,115 @@
+    @ObservedObject var viewModel: TripsViewModel
+    
+    @State private var selectedTimes: Set<TimeIntervalEnum> = []
+    @State private var showTransfers: Bool? = nil
```
`nil` можно не писать, если тип опциональный

NaR0RoW commented on TravelSchedule/TravelSchedule/View/WebView.swift:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    private func createURLRequest() -> URLRequest? {
+        return URLRequest(url: url)
```
В однострочных выражениях `return` можно опускать

NaR0RoW commented on TravelSchedule/TravelSchedule/ViewModel/CarriersViewModel.swift:
```diff
@@ -0,0 +1,24 @@
+        carriers = MockData.carriers
+    }
+    
+    func carrier(for name: String) -> CarrierModel? {
```
Метод подразумевает определенное действие. Рекомендую поменять название метода на:
`getCarrier`

NaR0RoW commented on TravelSchedule/TravelSchedule/ViewModel/CitiesViewModel.swift:
```diff
@@ -0,0 +1,40 @@
+        ]
+    }
+    
+    func city(for station: String) -> String {
```
Тут тоже лучше поменять на `getCity`

NaR0RoW commented on TravelSchedule/TravelSchedule/ViewModel/TripsViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+            let matchesTime = selectedTimes.contains { interval in
+                tripMatchesInterval(trip, interval: interval)
+            }
+            let matchesTransfer = showTransfers == nil || (showTransfers == false && trip.hasTransfers == false) || showTransfers == true
```
Если `Bool` значение не опционально, то его можно опустить:

``` swift
let matchesTransfer = showTransfers == nil || (showTransfers == false && !trip.hasTransfers) || showTransfers == true
```

UtinsStory committed:
Фикс после ревью

NaR0RoW commented on TravelSchedule/TravelSchedule/View/CarrierInfoView.swift:
```diff
@@ -8,14 +8,14 @@
     var body: some View {
         ZStack {
-            Color.ypWhite.edgesIgnoringSafeArea(.all)
+            Color.ypWhite.ignoresSafeArea(.all)
```
`.all` можно опустить, т.к. это аргумент по умолчанию

UtinsStory committed:
Попытка фикса NaN-value ошибки

UtinsStory committed:
Исправлена NaN-value ошибка. Добавлена проверка интернет-подключения при запросе городов с показом соответсвующего экрана при ошибке
// На данном этапе можно использовать моковые данные, поэтому показ ErrorView используется только в таком виде. В дальнейшем, при работе с сетью, показ ошибки буду использовать полноценно

NaR0RoW commented on TravelSchedule/TravelSchedule/Helpers/Reachability.swift:
```diff
@@ -0,0 +1,40 @@
+import Foundation
+import SystemConfiguration
+
+public class Reachability {
+    public func isConnectedToNetwork() -> Bool {
```
Модификатор доступа `public` можно убрать

NaR0RoW commented on TravelSchedule/TravelSchedule/Helpers/Reachability.swift:
```diff
@@ -0,0 +1,40 @@
+        let isReachable = flags.contains(.reachable)
+        let needsConnection = flags.contains(.connectionRequired)
+        
+        return (isReachable && !needsConnection)
```
Лучше не использовать круглые скобки, если в них нет явной необходимости

NaR0RoW commented on TravelSchedule/TravelSchedule/View/CityListView.swift:
```diff
@@ -12,49 +12,54 @@ struct CityListView: View {
     @StateObject private var viewModel = StationsViewModel()
     let selectAction: (String) -> Void
     @Binding var path: [Destination]
+    let reachability = Reachability()
```
Можно этот класс сделать синглтоном

NaR0RoW commented on TravelSchedule/TravelSchedule/View/CityListView.swift:
```diff
@@ -12,49 +12,54 @@ struct CityListView: View {
-                                Spacer()
-                                Image(systemName: "chevron.right")
-                                    .foregroundColor(.ypBlack)
+        if reachability.isConnectedToNetwork() == true {
```
`true` можно опустить:
`if reachability.isConnectedToNetwork() {`

NaR0RoW commented on TravelSchedule/TravelSchedule/View/ErrorView.swift:
```diff
@@ -8,31 +8,32 @@
 struct ErrorView: View {
-    var imageName: String
-    var message: String
+    @State var errorType: ErrorType
```
`@State` должен быть приватным

NaR0RoW commented on TravelSchedule/TravelSchedule/View/ErrorView.swift:
```diff
@@ -8,31 +8,32 @@
-            Text(message)
-                .font(.system(size: 24, weight: .bold))
-                .foregroundColor(.ypBlack)
+        VStack(alignment: .center, spacing: 16) {
```
`alignment: .center` можно опустить, т.к. это аргумент по умолчанию
